### ISSUE TITLE
Machine readable reserved and blocked domain lists

### ISSUE THREAD

**Issue by vohmar** [internetee/registry#1569]
Currently we publish the lists of reserved and blocked domains in a file on our homepage:

https://www.internet.ee/domains/special-conditions-for-registering-reserved-domains
https://www.internet.ee/domains/blocked-domains

As there is domain rules change in works that would enable releasing the domains currently in reserved list and set new principles on the content of those lists making these dynamic we should make the list available in machine readable and requestable form

REPP seems the preferable solution here

scope:
Enable registrars to request the list of reserved and blocked domains from registry. Option should also be available for EIS's homepage to be able to query and continue publishing the lists on the homepage.

data sources:
reserved list: db table reserved_domains (should be merged with disputes table at some point)
blocked list: db table blocked_domains

**Comment by maciej-szlosarczyk** [comment]
Changes that we agreed to implement yesterday, with one small change:

Add checksum to the response. It should be a SHA hash of all domain names got from the request. Stable ordering needs to be in place for the hashes to be reliable. Will allow registrars to only update the list if it's absolutely needed. Example:

On Monday, a new registrar GreatestDomains.com makes a request for the list, and gets a response back:

Request: `GET /repp/v1/retained_domains`

```ruby
{ 
  count: 120,
  checksum: "b6c9aee7acb1f40e96e7c89776dc2d58",
  domains: [ 
   # some long list here
  ],
  status: "new"
}
```

On Tuesday GreatestDomains.com makes a new request, but they kept the checksum on their side, so they send it as a query parameter:

Request: `GET /repp/v1/retained_domains?checksum=b6c9aee7acb1f40e96e7c89776dc2d58`

If the list has changed, they get similar response to the one above:

```ruby
{ 
  count: 118,
  checksum: "40676ca07b5c0b2f3960561c3ab74b17",
  domains: [ 
   # some long list here
  ],
  status: "new"
}
```

If it has not, no list gets sent over the wire:

```ruby
{ 
  checksum: "b6c9aee7acb1f40e96e7c89776dc2d58",
  status: "current"
}

Second change is to add `type` parameter so that the registrar can get list of only reserved domains separately. Example:

Request: `GET /repp/v1/retained_domains?type=reserved`

```ruby
{ 
  count: 60,
  checksum: "879a1cc86c662c10be176bae930a7a8c", # checksum for this list is naturally different than then one for a complete list.
  domains: [ 
   # some list of only reserved domains here
  ],
  status: "new"
}
```

Here comes one change from what we discussed yesterday @vohmar. I thought about the implications and adding combined filters (reserved and blocked, blocked and disputed etc) is something better to avoid. Parsing of lists in URL parameters is framework specific (Rails does it differently than let's say, AngularJS), so we would have to add POST method to be safe about it. As a result, the registrar, while integrating should know when to use POST and when to use GET. 

So my proposal here is to use only one filter at a time. If you want to get blocked and disputed domains, you need to do two requests.

And for auctions endpoint, I think this should be a separate ticket and that one should be straightforward: domains change daily, there is no need for caching, and the list is shorter than this one.

**Comment by maciej-szlosarczyk** [comment]
Actually this caching is going to be simpler, there's the ETag web standard and Rails seems to implement it. So a registrar makes initial request (here are examples with cURL):

`curl -v -X GET https://registry.test/repp/v1/retained_domains`

And the response looks like this:
```
HTTP/1.1 200 OK
Date: Fri, 15 May 2020 07:33:37 GMT
Content-Type: application/json; charset=utf-8
ETag: W/"a905b531243a6b0be42beb9d6ce60619"
Cache-Control: max-age=0, private, must-revalidate
X-Request-Id: e8a9855f-6dd1-443d-8720-dc26d5be3c9b
X-Runtime: 1.080339
Transfer-Encoding: chunked

{"count":1,"domains":[{"name":"reserved.test","status":"reserved","punycode_name":"reserved.test"}]}
```

Next request, the registrar should use the ETag to check if the resource has been changed

`$ curl -v -H "If-None-Match: W/\"a905b531243a6b0be42beb9d6ce60619\"" -X GET https://registry.test/repp/v1/retained_domains`

And the response has different HTTP status if the value has not changed:

```
HTTP/1.1 304 Not Modified
Date: Fri, 15 May 2020 07:36:08 GMT
ETag: W/"a905b531243a6b0be42beb9d6ce60619"
Cache-Control: max-age=0, private, must-revalidate
```

I will document it properly for registrars.

**Comment by vohmar** [comment]
closed by #1579
